lib/lacerda/compare/json_schema.rb

```diff
@@ -111,24 +87,127 @@ def schema_contains?(publish, consume, location = [])
 
         # We already know that publish and consume's type are equal
         # but if they're objects, we need to do some recursion
-        if consume['type'] == 'object'
-          consume['properties'].each do |property, schema|
-            return _e(:ERR_MISSING_PROPERTY, location, property) unless publish['properties'][property]
-            return false unless schema_contains?(publish['properties'][property], schema, location + [property])
+        if [consume['type']].flatten.include?('object')
+
+          # An object can either be described by its properties
+          # like this:
+          #
+          # (1) { "type": "object", "properties": { "active": { "type": "boolean" } }
+          #
+          # or by allowing a bunch of other types like this:
+          #
+          # (2) { "type": "object", "oneOf": [ {"$ref": "#/definitions/foo"}, {"type": "null"} ]
+          #
+          # So we need to take care of both cases for both "sides"
+          # (publish and consume), so 4 cases in total.
+          #
+          # First, the easy case:
+          if consume['properties'] and publish['properties']
+            consume['properties'].each do |property, schema|
+              return _e(:ERR_MISSING_PROPERTY, location, property) unless publish['properties'][property]
+              return false unless schema_contains?(publish: publish['properties'][property], consume: schema, location: location + [property])
+            end
+
+          # Now on to the trickier case, both have 'oneOf's:
+          #
+          # For each possible object type from the publish schema we have
+          # to check if we find a compatible type in the consume schema.
+          #
+          # It's not sufficient to just compare the names of the objects,
+          # because they might be different in the publish and consume
+          # schemas.
+          elsif publish['oneOf'] and consume['oneOf']
+            publish_types = publish['oneOf']
+            consume_types = [consume['oneOf']].flatten.compact
+
+            # Check all publish types for a compatible consume type
+            publish_types.each do |publish_type|
+              original_errors = @errors
+              @errors = []
+              compatible_consume_type_found = false
+              consume_types.each do |consume_type|

```

Comment:
It looks like you check all consume_types instead of stopping when you found one, is that on purpose?